### PR TITLE
drivers: serial: ns16550 uart driver init level based on parent node type

### PR DESCRIPTION
**Pull request by najumon1980**
add support for ns16550 uart driver with init level based on parent node type (PCI or non PCI).

### CODE REVIEW

File: drivers/serial/uart_ns16550.c [hunk 1]
```diff
@@ -1283,7 +1292,7 @@ static const struct uart_driver_api uart_ns16550_driver_api = {
 	};                                                                           \
 	DEVICE_DT_INST_DEFINE(n, &uart_ns16550_init, NULL,                           \
 			      &uart_ns16550_dev_data_##n, &uart_ns16550_dev_cfg_##n, \
-			      PRE_KERNEL_1, CONFIG_SERIAL_INIT_PRIORITY,             \
+			       BOOT_LEVEL(n), CONFIG_SERIAL_INIT_PRIORITY,             \
```
Comment (jhedberg) [inline drivers/serial/uart_ns16550.c:1295]: Looks like the alignment is different after this change - you've added one more space

File: boards/x86/rpl_crb/rpl_crb_defconfig [hunk 2]
```diff
@@ -13,3 +13,4 @@ CONFIG_X2APIC=y
 CONFIG_SMP=y
 CONFIG_BUILD_OUTPUT_EFI=y
 CONFIG_BUILD_NO_GAP_FILL=y
+CONFIG_UART_NS16550_AUTO_INIT_LEVEL=y
```
Comment (jhedberg) [inline boards/x86/rpl_crb/rpl_crb_defconfig:16]: What happens to other boards, like up_squared and ehl_crb, when your ACPI PR gets merged? Will they be broken?

File: drivers/serial/Kconfig.ns16550 [hunk 3]
```diff
@@ -72,6 +72,11 @@ config UART_NS16550_SIMULT_ACCESS
 	  When enabled, NS16550 supports IO, MMIO, PCIe UART devices simultaneously.
 	  For io-mapped instances, io-mapped DTS property need to be added in dtsi.
 
+config UART_NS16550_AUTO_INIT_LEVEL
```
Comment (dcpleung) [inline drivers/serial/Kconfig.ns16550:75]: This name is not what your commit is doing. It is not `AUTO`. You are changing the init level based on whether it is on PCI bus or not. Please rename it accordingly.

File: drivers/serial/uart_ns16550.c [hunk 4]
```diff
@@ -1248,6 +1248,15 @@ static const struct uart_driver_api uart_ns16550_driver_api = {
 #define DEV_DATA_DLF_INIT(n) \
 	_CONCAT(DEV_DATA_DLF, DT_INST_NODE_HAS_PROP(n, dlf))(n)
 
+#ifdef CONFIG_UART_NS16550_AUTO_INIT_LEVEL
+#define NS16550_BOOT_LEVEL0 PRE_KERNEL_1
+#define NS16550_BOOT_LEVEL1 POST_KERNEL
```
Comment (dcpleung) [inline drivers/serial/uart_ns16550.c:1253]: Why `POST_KERNEL`? As long as it is initialized after PCIe during `PRE_KERNEL_*`, it will still works.

**Comment by najumon1980** [comment]
Hi, could you please review and approve the PR (if ok) ?

File: drivers/serial/uart_ns16550.c (hunk 4 above)
Comment (dcpleung) [reply drivers/serial/uart_ns16550.c:1253]: TBH, I don't like this all-or-none approach. Could you a per device tree node approach, where each node would delay initialization based on some device tree property? (for example, add a `pcie-acpi` property to such relation.)

Comment (najumon1980) [reply drivers/serial/uart_ns16550.c:1253]: > TBH, I don't like this all-or-none approach. Could you a per device tree node approach, where each node would delay initialization based on some device tree property? (for example, add a `pcie-acpi` property to such relation.)

Current changes is per instance (node) and will check device node is on a pcie parent node or not.

**Review by dcpleung — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Comment by najumon1980** [comment]
A manual rebases done to resolve boards/x86/rpl_crb/rpl_crb_defconfig conflict with latest code.
 @jhedberg , @dcpleung could you please approve again

**Review by jhedberg — DISMISSED** [review]

File: drivers/serial/Kconfig.ns16550 [hunk 5]
```diff
@@ -72,6 +72,15 @@ config UART_NS16550_SIMULT_ACCESS
 	  When enabled, NS16550 supports IO, MMIO, PCIe UART devices simultaneously.
 	  For io-mapped instances, io-mapped DTS property need to be added in dtsi.
 
+config UART_NS16550_PARENT_INIT_LEVEL
+	bool "Boot level based on parent node"
```
Comment (jhedberg) [inline drivers/serial/Kconfig.ns16550:76]: Would it make sense to add `depends on ACPI` here?

Comment (jhedberg) [reply drivers/serial/Kconfig.ns16550:76]: Or perhaps `depends on ACPI && PCIE`?

Comment (najumon1980) [reply drivers/serial/Kconfig.ns16550:76]: better default y if ACPI, right?

Comment (jhedberg) [reply drivers/serial/Kconfig.ns16550:76]: > better default y if ACPI, right?

I guess that makes sense. And you can then drop the board-specific setting for RPL.

Comment (najumon1980) [reply drivers/serial/Kconfig.ns16550:76]: updated with default y if  ACPI enabled. Please review and approve if no more comments

**Review by jhedberg — APPROVED** [review]

**Comment by najumon1980** [comment]
@dcpleung, there was rebase done and also change in Kconfig for acpi. So kindly review and approve if ok.

**Review by dcpleung — APPROVED** [review]

**Pull request merged**